pkg/controller/cluster/main-controller.go

```diff
@@ -460,6 +500,34 @@ func (c *Controller) Start(threadiness int, stopCh <-chan struct{}) error {
 		select {}
 	}
 
+	// runSTS starts the STS API even if the pod is not the leader
+	runSTS := func(ctx context.Context) {
+		// stsServerWillStart is a channel for the STS Server API
+		stsServerWillStart := make(chan interface{})
+
+		go func() {
+			klog.Infof("Starting STS API server")
+			close(stsServerWillStart)
+			publicCertPath, publicKeyPath := c.waitSTSTLSCert()
+			certsManager, err := xcerts.NewManager(ctx, *publicCertPath, *publicKeyPath, LoadX509KeyPair)
+			if err != nil {
+				klog.Infof("STS HTTPS server ListenAndServeTLS failed: %v", err)
+				panic(err)
+			}
+			serverCertsManager = certsManager
+			c.sts.TLSConfig = c.createTLSConfig(serverCertsManager)
+			if err := c.sts.ListenAndServeTLS("", ""); err != http.ErrServerClosed {
+				klog.Infof("STS  HTTPS server ListenAndServeTLS failed: %v", err)
+				panic(err)
+			}
+		}()
+
+		klog.Info("Waiting for STS API to start")
+		<-stsServerWillStart
+
+		select {}
```

Comment:
this will cause goroutine leak, use return channel pattern, and exit the whole server gracefully if the `STS API Server` exists

Reply:
## return channel
```go
runSTS := func(ctx context.Context) <-chan interface{} {
 		// stsServerWillStart is a channel for the STS Server API
 		stsServerWillStart := make(chan interface{})

 		go func() {
 			klog.Infof("Starting STS API server")
 			close(stsServerWillStart)
 			publicCertPath, publicKeyPath := c.waitSTSTLSCert()
 			certsManager, err := xcerts.NewManager(ctx, *publicCertPath, *publicKeyPath, LoadX509KeyPair)
 			if err != nil {
 				klog.Infof("STS HTTPS server ListenAndServeTLS failed: %v", err)
 				panic(err)
 			}
 			serverCertsManager = certsManager
 			c.sts.TLSConfig = c.createTLSConfig(serverCertsManager)
 			if err := c.sts.ListenAndServeTLS("", ""); err != http.ErrServerClosed {
 				klog.Infof("STS  HTTPS server ListenAndServeTLS failed: %v", err)
 				panic(err)
 			}
 		}()

 		return stsServerWillStart
 	}
```

## block process
```go
if IsSTSEnabled() {
 		klog.Info("Waiting for STS API to start")
 		started := runSTS(ctx)
 		<-started
 	} else {
 		klog.Info("STS Api server is not enabled, not starting")
 	}

```